File: src/main/java/org/tasks/ui/DescriptionControlSet.java
package org.tasks.ui;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import com.google.common.base.Strings;
import com.todoroo.astrid.data.Task;

import org.tasks.R;
import org.tasks.injection.FragmentComponent;

import butterknife.BindView;
import butterknife.OnTextChanged;

public class DescriptionControlSet extends TaskEditControlFragment {

    public static final int TAG = R.string.TEA_ctrl_notes_pref;
    private static final String EXTRA_DESCRIPTION = "extra_description";

    @BindView(R.id.notes) EditText editText;

    private String description;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        if (savedInstanceState != null) {
            description = savedInstanceState.getString(EXTRA_DESCRIPTION);
        }
        if (!Strings.isNullOrEmpty(description)) {
            editText.setTextKeepState(description);
        }
        return view;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        outState.putString(EXTRA_DESCRIPTION, description);
    }

    @Override
    protected int getLayout() {
        return R.layout.control_set_description;
    }

    @Override
    protected int getIcon() {
        return R.drawable.ic_event_note_24dp;
    }

    @Override
    public int controlId() {
        return TAG;
    }

    @OnTextChanged(R.id.notes)
    void textChanged(CharSequence text) {
        description = text.toString().trim();
    }

    @Override
    public void initialize(boolean isNewTask, Task task) {
        description = task.getNotes();
    }

    @Override
    public void apply(Task task) {
        task.setNotes(description);
    }

    @Override
    public boolean hasChanges(Task original) {
        return !description.equals(original.getNotes());
    }

    @Override
    protected void inject(FragmentComponent component) {
        component.inject(this);
    }
}

File: app/src/main/java/org/tasks/ui/DescriptionControlSet.kt
package org.tasks.ui

import android.view.View
import android.view.ViewGroup
import androidx.compose.ui.platform.ComposeView
import com.google.android.material.composethemeadapter.MdcTheme
import dagger.hilt.android.AndroidEntryPoint
import org.tasks.R
import org.tasks.compose.edit.DescriptionRow
import org.tasks.dialogs.Linkify
import org.tasks.markdown.MarkdownProvider
import org.tasks.preferences.Preferences
import org.tasks.ui.TaskEditViewModel.Companion.stripCarriageReturns
import javax.inject.Inject

@AndroidEntryPoint
class DescriptionControlSet : TaskEditControlFragment() {
    @Inject lateinit var linkify: Linkify
    @Inject lateinit var preferences: Preferences
    @Inject lateinit var markdownProvider: MarkdownProvider

    private val linkifyEnabled: Boolean
        get() = preferences.getBoolean(R.string.p_linkify_task_edit, false)

    override fun bind(parent: ViewGroup?): View =
        (parent as ComposeView).apply {
            setContent {
                MdcTheme {
                    DescriptionRow(
                        text = viewModel.description.stripCarriageReturns(),
                        onChanged = { text -> viewModel.description = text.toString().trim { it <= ' ' } },
                        linkify = if (linkifyEnabled) linkify else null,
                        markdownProvider = markdownProvider,
                    )
                }
            }
        }

    override fun controlId() = TAG

    companion object {
        const val TAG = R.string.TEA_ctrl_notes_pref
    }
}
